src/black/files.py

```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
     """
     with open(path_config, "rb") as f:
         pyproject_toml = tomllib.load(f)
-    config = pyproject_toml.get("tool", {}).get("black", {})
-    return {k.replace("--", "").replace("-", "_"): v for k, v in config.items()}
+    config: Dict[str, Any] = pyproject_toml.get("tool", {}).get("black", {})
+    config = {k.replace("--", "").replace("-", "_"): v for k, v in config.items()}
+
+    if "target_version" not in config:
+        inferred_target_version = infer_target_version(pyproject_toml)
+        if inferred_target_version is not None:
+            config["target_version"] = [inferred_target_version.name.lower()]
+
+    return config
+
+
+def infer_target_version(pyproject_toml: Dict[str, Any]) -> Optional[TargetVersion]:
+    """Infer Black's target version from the project metadata in pyproject.toml.
+
+    If target version cannot be inferred, returns None.
+    """
+    requires_python = get_req_python(pyproject_toml)
+    if requires_python is not None:
+        return parse_req_python(requires_python)
+    return None
+
+
+def get_req_python(pyproject_toml: Dict[str, Any]) -> Optional[str]:
+    """Get the required Python version from the project metadata.
+
+    Currently only supports the PyPA standard format:
+    https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#requires-python
+
+    If the field is not present or cannot be parsed, returns None.
+    """
+    project_metadata = pyproject_toml.get("project", {})
+    requires_python: Optional[str] = project_metadata.get("requires-python", None)
+    return requires_python
+
+
+def parse_req_python(requires_python: str) -> Optional[TargetVersion]:
+    """Parse the requires-python field to determine Black's target version.
+
+    If the field cannot be parsed, returns None.
+    """
+    try:
+        return parse_req_python_version(requires_python)
+    except InvalidVersion:
+        pass
+    try:
+        return parse_req_python_specifier(requires_python)
+    except InvalidSpecifier:
+        pass
+    return None
+
+
+def parse_req_python_version(requires_python: str) -> Optional[TargetVersion]:
+    """Parse a version string (i.e. ``"3.7"``) to a TargetVersion.
+
+    If parsing fails, will raise a packaging.version.InvalidVersion error.
+    If the parsed version cannot be mapped to a valid TargetVersion, returns None.
+    """
+    version = Version(requires_python)
+    try:
+        return TargetVersion(version.release[1])
+    except (IndexError, ValueError):
+        return None
+
+
+def parse_req_python_specifier(requires_python: str) -> Optional[TargetVersion]:
+    """Parse a specifier string (i.e. ``">=3.7,<3.10"``) to a TargetVersion.
+
+    If parsing fails, will raise a packaging.specifiers.InvalidSpecifier error.
+    If the parsed specifier cannot be mapped to a valid TargetVersion, returns None.
+    """
+    specifier_set = strip_specifier_set(SpecifierSet(requires_python))
```

Comment:
Black will blow up with `Error: Could not open file '/home/ichard26/programming/oss/black/pyproject.toml': Error reading configuration file: Invalid version: 'a'` if you pass a specifier set that has an invalid version that's sanitized (eg. `>=3.7,!=a`).

Given the defensive style of the rest of the logic, it's probably best to return None if `InvalidVersion` is raised here. The other option is to only ignore invalid specifiers by catching `InvalidVersion` in `strip_specifier_set`'s loop.

Reply:
Good catch - let's indeed return None in this case. I will figure out how to do this nicely and add a commit for this.

Reply:
This is now fixed!